server/src/main/scala/com/cloudera/livy/utils/SparkYarnApp.scala

```diff
@@ -0,0 +1,230 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.cloudera.livy.utils
+
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeoutException
+
+import scala.annotation.tailrec
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent._
+import scala.concurrent.duration._
+import scala.language.postfixOps
+
+import org.apache.hadoop.yarn.api.records.{ApplicationId, ApplicationReport, FinalApplicationStatus, YarnApplicationState}
+import org.apache.hadoop.yarn.client.api.YarnClient
+import org.apache.hadoop.yarn.conf.YarnConfiguration
+
+import com.cloudera.livy.{LivyConf, Logging, Utils}
+import com.cloudera.livy.rsc.{Utils => RSCUtils}
+import com.cloudera.livy.util.LineBufferedProcess
+
+object SparkYarnApp extends Logging {
+  // Create a non daemon thread pool to run getAppIdFromTag().
+  // getAppIdFromTag() might take a while and should not use the global default thread pool.
+  private implicit val ec = ExecutionContext.fromExecutor(
+    Executors.newCachedThreadPool(RSCUtils.newDaemonThreadFactory("getAppIdFromTag-%d")))
+
+  // YarnClient is thread safe. Create once, share it across threads.
+  lazy val yarnClient = {
+    val c = YarnClient.createYarnClient()
+    c.init(new YarnConfiguration())
+    c.start()
+    c
+  }
+
+  def fromAppTag(
+      appTag: String,
+      process: Option[LineBufferedProcess],
+      listener: Option[SparkAppListener],
+      livyConf: LivyConf): SparkYarnApp = {
+    val deadline = getYarnTagToAppIdTimeout(livyConf).fromNow
+    val pollInterval = getYarnPollInterval(livyConf)
+    new SparkYarnApp(getAppIdFromTag(appTag, pollInterval, deadline), process, listener, livyConf)
+  }
+
+  /**
+   * Find the corresponding YARN application id from an application tag.
+   *
+   * @param appTag The application tag tagged on the target application.
+   *               If the tag is not unique, it returns the first application it found.
+   *               It will be converted to lower case to match YARN's behaviour.
+   * @return ApplicationId or the failure.
+   */
+  @tailrec
+  private def getAppIdFromTag(
+      appTag: String,
+      pollInterval: Duration,
+      deadline: Deadline): ApplicationId = {
+    val appTagLowerCase = appTag.toLowerCase()
+
+    // FIXME Should not loop thru all YARN applications but YarnClient doesn't offer an API.
+    // Consider calling rmClient in YarnClient directly.
+    val appType = Set("SPARK").asJava
+    yarnClient.getApplications(appType).asScala.find(_.getApplicationTags.contains(appTagLowerCase))
+    match {
+      case Some(app) => app.getApplicationId
+      case None =>
+        if (deadline.isOverdue) {
+          throw new Exception(s"No YARN application is tagged with $appTagLowerCase.")
+        } else {
+          Clock.sleep(pollInterval.toMillis)
```

Comment:
`SparkYarnApp` has code to wait for a process when a process does not exist in the recovery case. The "process does not exist" is handled when matching for a process result, which is weird.

The actual sequence of operations is "wait for process -> match tag to app id -> monitor app through yarn", and that logic is scattered all over the place. For example, the "wait for process" part is tightly coupled with the code that matches tags to app ids. It's a simple change to decouple those, and that would make the code clearer. No process? Skip any process-related code. Process fails? Throw an exception. Any other state? Ok, *now* do tag matching.

Maybe two classes is overkill, but the current code really needs some rearranging so that it's easier to follow the flow of operations.